src/pip/_internal/build_env.py

```diff
@@ -29,6 +28,29 @@
 
 logger = logging.getLogger(__name__)
 
+PIP_RUNNER = """
+import importlib.util
+import os
+import runpy
+import sys
+
+
+class PipImportRedirectingFinder:
+
+    @classmethod
+    def find_spec(cls, fullname, path=None, target=None):
+        if not fullname.startswith("pip."):
+            return None
+
+        # Import pip from the current source directory
+        location = os.path.join({source!r}, *fullname.split("."))
+        return importlib.util.spec_from_file_location(fullname, location)
+
+
+sys.meta_path.append(PipImportRedirectingFinder())
```

Comment:
Is it worth inserting our finder at the *start* of `sys.meta_path`, to avoid any other hooks getting called first? Othewise, this LGTM.